Add TF DeiT implementation

amyeroberts opened the pull request:
# What does this PR do?

Adds the TF implementations of DeiT

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

amyeroberts committed:
Initial TF DeiT implementation

amyeroberts committed:
Initial TF DeiT implementation

amyeroberts committed:
Fix copies naming issues

amyeroberts committed:
Fix copies naming issues

amyeroberts committed:
Fix up + docs

amyeroberts committed:
Fix up + docs

amyeroberts committed:
Properly same main layer

amyeroberts committed:
Properly same main layer

amyeroberts committed:
Name layers properly

amyeroberts committed:
Name layers properly

amyeroberts committed:
Merge branch 'add-tf-deit' of github.com:amyeroberts/transformers into add-tf-deit

amyeroberts committed:
Fixup

amyeroberts committed:
Fix import

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_17806). All of your documentation changes will be reflected on that endpoint.

amyeroberts committed:
Fix import

amyeroberts committed:
Fix import

amyeroberts committed:
Fix weight loading for tests whilst not on hub

amyeroberts committed:
Add doc tests and remove to_2tuple

amyeroberts committed:
Add back to_2tuple
Removing to_2tuple results in many downstream changes needed because of the copies checks

NielsRogge commented:
Please also incorporate the updates of #17731

amyeroberts commented:
> Please also incorporate the updates of https://github.com/huggingface/transformers/pull/17731

@NielsRogge Will do! I originally added, but it resulted in lots of changes because of all the `Copied From` statements, so will wait until your PR is merged and those updates finalised.

amyeroberts committed:
Merge in main and resolve conflicts

amyeroberts committed:
Incorporate updates in Improve vision models #17731 PR

amyeroberts committed:
Merge branch 'main' into add-tf-deit

amyeroberts committed:
Merge branch 'main' into add-tf-deit

amyeroberts committed:
Don't hard code num_channels

amyeroberts committed:
Copy PyTorch DeiT embeddings and remove pytorch operations with mask

amyeroberts committed:
Fix patch embeddings & tidy up

amyeroberts committed:
Merge branch 'main' into add-tf-deit

amyeroberts committed:
Update PixelShuffle to move logic into class layer

amyeroberts committed:
Update doc strings - remove PT references

amyeroberts committed:
Use NHWC format in internal layers

amyeroberts committed:
Resolve conflict - two models added to auto

amyeroberts committed:
Fix up

amyeroberts committed:
Merge branch 'main' into add-tf-deit

amyeroberts committed:
Use linear activation layer

amyeroberts committed:
Remove unused import

amyeroberts commented on tests/models/deit/test_modeling_tf_deit.py:
```diff
@@ -0,0 +1,288 @@
+    @slow
+    def test_inference_image_classification_head(self):
+        model = TFDeiTForImageClassificationWithTeacher.from_pretrained(
+            "facebook/deit-base-distilled-patch16-224", from_pt=True
```
To be removed after weights pushed to hub

amyeroberts commented on src/transformers/models/deit/modeling_tf_deit.py:
```diff
@@ -0,0 +1,1047 @@
+        >>> # so the head will be randomly initialized, hence the predictions will be random
+        >>> feature_extractor = DeiTFeatureExtractor.from_pretrained("facebook/deit-base-distilled-patch16-224")
+        >>> model = TFDeiTForImageClassification.from_pretrained(
+        ...     "facebook/deit-base-distilled-patch16-224", from_pt=True
```
To be removed after weights pushed to hub

amyeroberts commented on tests/models/deit/test_modeling_tf_deit.py:
```diff
@@ -0,0 +1,288 @@
+    @slow
+    def test_model_from_pretrained(self):
+        for model_name in TF_DEIT_PRETRAINED_MODEL_ARCHIVE_LIST[:1]:
+            model = TFDeiTModel.from_pretrained(model_name, from_pt=True)
```
To be removed after weights pushed to hub

Rocketknight1 reviewed (approved):
I couldn't see any issues with the ported TF code anywhere! (Except my usual nitpicks that people are using `tf.multiply` and `tf.matmul` instead of just `*` and `@`, but that's purely cosmetic and copied from another file anyway)

Also, tests seem comprehensive, so this PR looks good to me!

patrickvonplaten reviewed (commented):
Deferring to @NielsRogge and @gante here as it's Vision and TF. Happy to do a final review though if you want :-)

NielsRogge commented on src/transformers/models/deit/modeling_tf_deit.py:
```diff
@@ -0,0 +1,1047 @@
+# coding=utf-8
+# Copyright 2021 Facebook AI Research (FAIR), Ross Wightman, The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2022 Facebook AI Research (FAIR), Ross Wightman, The HuggingFace Inc. team. All rights reserved.
```

NielsRogge commented on src/transformers/models/deit/modeling_tf_deit.py:
```diff
@@ -0,0 +1,1047 @@
+        )
+
+
+@dataclass
```
Usually I add model output classes at the top of the modeling file

amyeroberts replied:
Agreed - it's much nicer. I've moved it to the top :)

NielsRogge reviewed (approved):
LGTM! Very clean!

sgugger commented on tests/models/deit/test_modeling_tf_deit.py:
```diff
@@ -0,0 +1,288 @@
+
+    def prepare_config_and_inputs_for_common(self):
+        config_and_inputs = self.prepare_config_and_inputs()
+        (
+            config,
+            pixel_values,
+            labels,
+        ) = config_and_inputs
```
```suggestion
        config, pixel_values, labels = config_and_inputs
```

sgugger reviewed (approved):
Very clean, thanks a lot for porting this model to TensorFlow!

amyeroberts committed:
Apply suggestions from code review

Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>
Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>

Co-authored-by: NielsRogge <48327001+NielsRogge@users.noreply.github.com>
Co-authored-by: Sylvain Gugger <35901082+sgugger@users.noreply.github.com>

amyeroberts committed:
Move dataclass to top of file

amyeroberts committed:
Remove from_pt now weights on hub

amyeroberts committed:
Merge branch 'main' into add-tf-deit

amyeroberts committed:
Fixup
